File: LC/[0300] Longest increasing Subsequence - Binary Search.py
from bisect import bisect_left
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        arr=[]

        for nn in nums:
            idx=bisect_left(arr, nn)
            if idx >= len(arr):
                arr.append(nn)
            else:
                arr[idx]=nn

        return len(arr)

File: LC/[0300] Longest increasing Subsequence - Binary Search.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        List<Integer> list = new ArrayList<>();
        for(var nn: nums){
            int idx = Collections.binarySearch(list, nn);
            if(idx<0) idx = -idx-1;
            if(idx >= list.size()){
                list.add(nn);
            }else{
                list.set(idx, nn);
            }
        }
        return list.size();
    }
}

    // int idx = Collections.binarySearch(list, x);